src/utils/ScreenReader.ts

```diff
@@ -0,0 +1,281 @@
+
+  /**
+   * Stops any speaking that is currently happening.
+   */
+  stop(): void
+}
+
+/**
+ * Implement this to provide screen reading.
+ */
+export interface ScreenReader {
+  /**
+   * Call this with an event target when a focus event occurs. Resolves when speaking is done.
+   */
+  onFocus(target?: EventTarget): Promise<void>
+
+  /**
+   * Call this with an event target when a click event occurs. Resolves when speaking is done.
+   */
+  onClick(target?: EventTarget): Promise<void>
+
+  /**
+   * Call this when a page load occurs. Resolves when speaking is done.
+   */
+  onPageLoad(): Promise<void>
+
+  /**
+   * Call this when the screen reader is enabled. Resolves when speaking is done.
+   */
+  onScreenReaderEnabled(): Promise<void>
+
+  /**
+   * Call this when the screen reader is disabled. Resolves when speaking is done.
+   */
+  onScreenReaderDisabled(): Promise<void>
+
+  /**
+   * Directly triggers speech of text. Resolves when speaking is done.
+   */
+  speak(text: string, options?: SpeakOptions): Promise<void>
+
+  /**
+   * Directly triggers speech of an element. Resolves when speaking is done.
+   */
+  speakNode(element: Element, options?: SpeakOptions): Promise<void>
+
+  /**
+   * Directly triggers speech of an event target. Resolves when speaking is done.
+   */
+  speakEventTarget(target?: EventTarget, options?: SpeakOptions): Promise<void>
+}
+
+/**
+ * Implements `ScreenReader` using the ARIA DOM attributes.
+ */
+export class AriaScreenReader implements ScreenReader {
+  private readonly tts: TextToSpeech
+
+  /**
+   * @param tts A text-to-speech engine to use to speak aloud.
+   */
+  public constructor(tts: TextToSpeech) {
+    this.tts = tts
+  }
+
+  /**
+   * Call this with an event target when a focus event occurs. Resolves when speaking is done.
+   */
+  public onFocus = async (target?: EventTarget) => {
+    await this.speakEventTarget(target)
+  }
+
+  /**
+   * Call this with an event target when a click event occurs. Resolves when speaking is done.
+   */
+  public onClick = async (target?: EventTarget) => {
+    await this.speakEventTarget(target)
+  }
+
+  /**
+   * Call this when a page load occurs. Resolves when speaking is done.
+   */
+  public async onPageLoad(): Promise<void> {
+    this.tts.stop()
+  }
+
+  /**
+   * Call this when the screen reader is enabled. Resolves when speaking is done.
+   */
+  public async onScreenReaderEnabled(): Promise<void> {
+    await this.speak('Screen reader enabled', { now: true })
+  }
+
+  /**
+   * Call this when the screen reader is disabled. Resolves when speaking is done.
+   */
+  public async onScreenReaderDisabled(): Promise<void> {
+    await this.speak('Screen reader disabled', { now: true })
+  }
+
+  /**
+   * Directly triggers speech of text. Resolves when speaking is done.
+   */
+  public async speak(text: string, options: SpeakOptions = {}): Promise<void> {
+    /* istanbul ignore next */
+    if (process.env.NODE_ENV === 'development') {
```

Comment:
This is here so we don't get a bunch of junk when the tests run, but it's useful to see in development.